lib/rubygems/installer.rb

```diff
@@ -953,13 +953,7 @@ def write_cache_file
   end
 
   def ensure_writable_dir(dir) # :nodoc:
-    begin
-      Dir.mkdir dir, *[options[:dir_mode] && 0o755].compact
-    rescue SystemCallError
-      raise unless File.directory? dir
-    end
-
-    raise Gem::FilePermissionError.new(dir) unless File.writable? dir
```

Comment:
I think test failures are because of this line getting removed. I meant to simplify the `begin .. rescue SystemCallError .. end` block to just the `FileUtils.mkdir_p` line but I think this line is still necessary.

That's why I also suggested to modify the Bundler override as in https://github.com/rubygems/rubygems/pull/8755#pullrequestreview-2910900718, since form `Bundler`'s point of view we don't need the specialized error since there it can actually be simplified to the implementation you're proposing here.